File: src/main/java/net/kenro/ji/jin/purescript/highlighting/PSSyntaxHighlightAnnotator.java
package net.kenro.ji.jin.purescript.highlighting;

import com.intellij.lang.annotation.Annotation;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.psi.PsiElement;
import net.kenro.ji.jin.purescript.psi.PSElements;
import org.jetbrains.annotations.NotNull;

import static com.intellij.patterns.PlatformPatterns.psiElement;


public class PSSyntaxHighlightAnnotator implements Annotator {
    @Override
    public void annotate(@NotNull final PsiElement element, @NotNull final AnnotationHolder holder) {
        if (psiElement(PSElements.ValueRef).accepts(element)) {
            final String text = element.getText();
            holder.newAnnotation(HighlightSeverity.INFORMATION, text)
                .textAttributes(PSSyntaxHighlighter.IMPORT_REF)
                .create();
        } else if (psiElement(PSElements.TypeAnnotationName).accepts(element)) {
            holder.newAnnotation(HighlightSeverity.INFORMATION, element.getText())
                .textAttributes(PSSyntaxHighlighter.TYPE_ANNOTATION_NAME)
                .create();
        } else if ((psiElement(PSElements.PositionedDeclarationRef).accepts(element)
                || psiElement(PSElements.TypeConstructor).accepts(element)
                || psiElement(PSElements.pClassName).accepts(element))) {
//                || psiElement(PSElements.pModuleName).accepts(element))) {
            holder
                .newAnnotation(HighlightSeverity.INFORMATION, element.getText())
                .textAttributes(PSSyntaxHighlighter.TYPE_NAME)
                .create();
        } else if ((psiElement(PSElements.GenericIdentifier).accepts(element)
                || psiElement(PSElements.Constructor).accepts(element)
                || psiElement(PSElements.qualifiedModuleName).accepts(element))) {
            holder.newAnnotation(HighlightSeverity.INFORMATION, element.getText())
                .textAttributes(PSSyntaxHighlighter.TYPE_VARIABLE)
                .create();
        } else if (psiElement(PSElements.LocalIdentifier).accepts(element)) {
            holder.newAnnotation(HighlightSeverity.INFORMATION, element.getText())
                .textAttributes(PSSyntaxHighlighter.NUMBER)
                .create();
        }
    }


}

File: src/main/java/org/purescript/highlighting/PSSyntaxHighlightAnnotator.kt
package org.purescript.highlighting

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
import com.intellij.psi.PsiElement
import org.purescript.highlighting.PSSyntaxHighlighter.Companion.IMPORT_REF
import org.purescript.highlighting.PSSyntaxHighlighter.Companion.NUMBER
import org.purescript.highlighting.PSSyntaxHighlighter.Companion.TYPE_ANNOTATION_NAME
import org.purescript.highlighting.PSSyntaxHighlighter.Companion.TYPE_NAME
import org.purescript.highlighting.PSSyntaxHighlighter.Companion.TYPE_VARIABLE
import org.purescript.parser.PSElements.Companion.Constructor
import org.purescript.parser.PSElements.Companion.GenericIdentifier
import org.purescript.parser.PSElements.Companion.LocalIdentifier
import org.purescript.parser.PSElements.Companion.PositionedDeclarationRef
import org.purescript.parser.PSElements.Companion.TypeAnnotationName
import org.purescript.parser.PSElements.Companion.TypeConstructor
import org.purescript.parser.PSElements.Companion.ValueRef
import org.purescript.parser.PSElements.Companion.pClassName

class PSSyntaxHighlightAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        when (element.node.elementType) {
            ValueRef -> {
                holder.newSilentAnnotation(INFORMATION)
                    .textAttributes(IMPORT_REF)
                    .create()
            }
            TypeAnnotationName -> {
                holder.newSilentAnnotation(INFORMATION)
                    .textAttributes(TYPE_ANNOTATION_NAME)
                    .create()
            }
            PositionedDeclarationRef, TypeConstructor, pClassName -> {
                holder.newSilentAnnotation(INFORMATION)
                    .textAttributes(TYPE_NAME)
                    .create()
            }
            GenericIdentifier, Constructor -> {
                holder.newSilentAnnotation(INFORMATION)
                    .textAttributes(TYPE_VARIABLE)
                    .create()
            }
            LocalIdentifier -> {
                holder.newSilentAnnotation(INFORMATION)
                    .textAttributes(NUMBER)
                    .create()
            }
        }
    }
}